### PR TITLE
Add and use `generics.is_empty()`, rather than checking generics' params vec

### PR DESCRIPTION
**Pull request by compiler-errors**
r? spastorino

### CODE REVIEW

File: compiler/rustc_hir_analysis/src/check/wfcheck.rs [hunk 1]
```diff
@@ -430,7 +430,7 @@ fn check_gat_where_clauses(tcx: TyCtxt<'_>, trait_def_id: LocalDefId) {
             }
             let gat_generics = tcx.generics_of(gat_def_id);
             // FIXME(jackh726): we can also warn in the more general case
-            if gat_generics.params.is_empty() {
+            if gat_generics.is_empty() {
```
Comment (compiler-errors) [inline compiler/rustc_hir_analysis/src/check/wfcheck.rs:433]: Oops, this is wrong!

File: compiler/rustc_hir_typeck/src/method/confirm.rs [hunk 2]
```diff
@@ -450,7 +450,7 @@ impl<'a, 'tcx> ConfirmContext<'a, 'tcx> {
         // `foo.bar::<u32>(...)` -- the `Self` type here will be the
         // type of `foo` (possibly adjusted), but we don't want to
         // include that. We want just the `[_, u32]` part.
-        if !args.is_empty() && !generics.params.is_empty() {
+        if !args.is_empty() && !generics.is_empty() {
```
Comment (compiler-errors) [inline compiler/rustc_hir_typeck/src/method/confirm.rs:453]: This is also wrong

File: compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs [hunk 3]
```diff
@@ -411,7 +411,7 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
         // Traits always have `Self` as a generic parameter, which means they will not return early
         // here and so associated type bindings will be handled regardless of whether there are any
         // non-`Self` generic parameters.
-        if generics.params.is_empty() {
+        if generics.is_empty() {
```
Comment (compiler-errors) [inline compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs:414]: This is wrong

File: compiler/rustc_hir_typeck/src/method/prelude2021.rs [hunk 4]
```diff
@@ -282,7 +282,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 if !self_ty_name.contains('<') {
                     if let Adt(def, _) = self_ty.kind() {
                         let generics = self.tcx.generics_of(def.did());
-                        if !generics.params.is_empty() {
+                        if !generics.is_empty() {
```
Comment (compiler-errors) [inline compiler/rustc_hir_typeck/src/method/prelude2021.rs:285]: This is wrong

File: compiler/rustc_hir_typeck/src/method/probe.rs [hunk 5]
```diff
@@ -1871,7 +1871,7 @@ impl<'a, 'tcx> ProbeContext<'a, 'tcx> {
         let generics = self.tcx.generics_of(method);
         assert_eq!(args.len(), generics.parent_count);
 
-        let xform_fn_sig = if generics.params.is_empty() {
+        let xform_fn_sig = if generics.is_empty() {
```
Comment (compiler-errors) [inline compiler/rustc_hir_typeck/src/method/probe.rs:1874]: This is wrong

File: compiler/rustc_middle/src/ty/print/mod.rs [hunk 6]
```diff
@@ -157,7 +157,7 @@ pub trait Printer<'tcx>: Sized {
                         // If we have any generic arguments to print, we do that
                         // on top of the same path, but without its own generics.
                         _ => {
-                            if !generics.params.is_empty() && args.len() >= generics.count() {
+                            if !generics.is_empty() && args.len() >= generics.count() {
```
Comment (compiler-errors) [inline compiler/rustc_middle/src/ty/print/mod.rs:160]: This is wrong

File: compiler/rustc_trait_selection/src/traits/object_safety.rs [hunk 7]
```diff
@@ -397,7 +397,7 @@ pub fn object_safety_violations_for_assoc_item(
         // Associated types can only be object safe if they have `Self: Sized` bounds.
         ty::AssocKind::Type => {
             if !tcx.features().generic_associated_types_extended
-                && !tcx.generics_of(item.def_id).params.is_empty()
+                && !tcx.generics_of(item.def_id).is_empty()
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/object_safety.rs:400]: This is wrong

File: compiler/rustc_trait_selection/src/traits/select/mod.rs [hunk 8]
```diff
@@ -1773,7 +1773,7 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
             // If this type is a GAT, and of the GAT args resolve to something new,
             // that means that we must have newly inferred something about the GAT.
             // We should give up in that case.
-            if !generics.params.is_empty()
+            if !generics.is_empty()
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/select/mod.rs:1776]: This is wrong

File: src/tools/clippy/clippy_lints/src/iter_without_into_iter.rs [hunk 9]
```diff
@@ -225,7 +225,7 @@ impl {self_ty_without_ref} {{
             && let ImplItemKind::Fn(sig, _) = item.kind
             && let FnRetTy::Return(ret) = sig.decl.output
             && is_nameable_in_impl_trait(ret)
-            && cx.tcx.generics_of(item_did).params.is_empty()
+            && cx.tcx.generics_of(item_did).is_empty()
```
Comment (compiler-errors) [inline src/tools/clippy/clippy_lints/src/iter_without_into_iter.rs:228]: Hm, this is wrong

**Review by compiler-errors — COMMENTED** [review]
Yeah, after a self-review it's kind of certain there are more places where it's wrong to use this than right.

**Comment by spastorino** [comment]
I guess you have decided to not proceed with this change, right?. I was suggesting something like an `is_empty` and an `is_own_empty` or something like that where you have both functions and we stop doing `.params.is_empty` and/or using `count`.

**Comment by compiler-errors** [comment]
@spastorino: feel free to do the change yourself -- you can use the comments above to figure out where we care about is_own_empty vs is_empty.

**Pull request closed without merging**